xc/device.ml

```diff
@@ -1493,7 +1493,13 @@ module Dm = struct
 
     let qmp_event_handle domid qmp_event =
       (* This function will be extended to handle qmp events *)
-      debug "Got QMP event, domain-%d: %s" domid qmp_event.event
+      debug "Got QMP event, domain-%d: %s" domid qmp_event.event;
+      qmp_event.data >>= fun d ->
+      match d with
+      | RTC_CHANGE timeoffset ->
+        with_xs (fun xs ->
+          xs.Xs.write ("/vm/" ^ (Uuidm.to_string (Xenops_helpers.uuid_of_domid ~xs domid)) ^ "/rtc/timeoffset") (string_of_int timeoffset)
```

Comment:
Writing to xenstore is a rather indirect way of achieving this -- I'll leave it to others to decide if that is best.

Reply:
Even for `qemu-trad`, I think this is a long way to update `VM.platform[timeoffset]`
1. qemu-trad writes timeoffset value to xenstore, the watcher thread adds `Dynamic.Vm` item to the Updates queue 
https://github.com/xapi-project/xenopsd/blob/master/xc/xenops_server_xen.ml#L2967
2. The updates thread executes `VM_check_state` -> `B.VM.get_state`, which gets timeoffset from xenstore, then sends vm state to xen-api process by Unix socket.
https://github.com/xapi-project/xenopsd/blob/master/xc/xenops_server_xen.ml#L1658
3. xen-api process receives xenops message and updates `VM.platform[timeoffset]`
https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/xapi_xenops.ml#L1493

So, for `QEMU-upstream`, I think we use the existing mechanism, just write the value to xenstore.
Hi @mg12 , what do you think?

Reply:
you kept the existing mechanism present in qemu-trad, which is good for preventing regressions, but I wonder if @jonludlam can think of a better way which we should use to implement this for upstream-qemu.